README.markdown
Alligator is a simple application server built on top of AntiNode and Node.js 

Latest Version 0.37

# Usage

Run it from the command line. 

   $ node alligator.js 

Requires Node.JS v0.1.99 or greater.
If you want to bind to a port under 1024, you'll need to run node with special
privileges.




# Features
- Handling request parameters easily (for GET and POST)
- Distributed session management, including timeout
- Distributed application scope management
- Dynamic js loader (the lib folder)
- User can set the server-side script begin and end tags
- Utilizing all cores per cpu (the `nodes` settings parameter)

When one is writing a script nested to the <? some;JavaScript;Code;In Here;?> tags

He/She can use the following implicit functions/variables over the context object:

- `context.request` for handling the request
- `context.request.parameters` to read and write request post/get parameters
- `context.responseHead.header` to add headers of the response (e.g. responseHead.header["set-cookie"] = "..";)
- `context.responseHead.status` to change the HTTP response status
- `context.write(str)` to add string to the response body (or instead one can use the <?=str?> tag)
- `context.forward(other.jssp)` to forward the request and response to another server-side resource
- `context.sendRedirect(url)` to send HTTP redirect response back to the client
- `context.lib.filename.member` to access whatever lib one has loaded 
- `context.session.set(key,value,callbackFunc)` to put anything on the HttpSession
- `context.session.get(key,callbackFunc)` to get anything from the HttpSession
- `context.application.set(key,value,callbackFunc)` to put anything on the application context
- `context.application.get(key,callbackFunc)` to get anything from the application context



#Examples
Ex1.jssp:
	<? var a = 1+1;?><br/>
	<? context.write(a);?>
Translates to:
	<br/>2
Once can achieve the same thing using Ex2.jssp
	<? var a = 1+1;?><br/>
	<?=a?>

One can forward from one jssp to another:
Exf1.jssp:
	<? var dbInfo= gettingInfoFromDatabase();
	   context.request.parameters.db = dbInfo;
           context.forward("showTable.jssp");?>
showTable.jssp:
	<?=genetrateHTMLTable(request.parameters.db)?>

One can easily redirect:
	<? if(request.parameters.googleIt=="true")
		context.sendRedirect("http://www.google.com");
	else{?>
	<H1> Welcome...</H1>
	<?}?>

How to use the session scope, the counter Example: (separation of logic and view)
logic.jssp:
	<?
		var counter = 1;
		context.session.get("counter",function(value){
			context.log.debug("Page","SESSIONLOGIC.JSSP, value - " +value);
			if(value == undefined){
				context.session.set("counter",1);
			}else{
				counter = value+1;
				context.session.set("counter",counter);
			}
			context.request.parameters.counter = counter;
			context.forward("counter/view.jssp");				
		});				
	?>
view.jssp:
	<HTML>
		<HEAD><TITLE>Application Scope Counter Tester</TITLE></HEAD>
		<BODY>
		<?
			var counter = request.parameters.counter;
			if(counter==1)
				context.write("First Time");
			else
				context.write("Number of hits by you:" + counter);
				
		?>
		</BODY>
	</HTML>
	
In case you had like to write JS only you can use a file with the 'script_ext', here is a different way to write logic.jssp
logic.ssjs: (no script tags)
	this.page = function (context){
		var counter = 1;
		context.application.get("counter",function(value){
			context.log.debug("Page","ApplicationLOGIC.JSSP, value - " +value);
			if(value == undefined){
				context.application.set("counter",1);
			}else{
				counter = value+1;
				context.application.set("counter",counter);
			}
			context.request.parameters.counter = counter;
			context.forward("counter/view.jssp");				
		});					
	
	}			


**Please note that behind the scene Alligator translate JSSP file to SSJS file upon the first request to this JSSP file(lazy)
and later on use the SSJS file only.
One can develop SSJS files instead of JSSP which makes the translation unnecessary.
Alligator invokes the page() function.

One can also add content outside the page() function in two ways:
using the <?! code here ?> tag
sample.jssp:
	<?!
		this.test = function(context){
			context.write("hello ajax world");
		};
	?>
	
	<html>
		<head>
			<script src="jquery142.js"></script>
			<script src="alligator.js"></script>
			<script>
				function click1(){
					server("test",function(text){
						$("#hello").html(text);
					});
				}
			</script>
		</head>
		<body>
			<span id="hello"></span><br/>
			 <form>
	 			<input type="button" id="button1" value="click me" onclick="click1()" />
	 		</form>
		</body>
	</html>
which compiles to sample.ssjs:(the text is escaped)
	this.page = function (context){context.writeEscapedText("");
		context.writeEscapedText("%3Chtml%3E%0A");
		context.writeEscapedText("%09%3Chead%3E%0A");
		context.writeEscapedText("%09%09%3Cscript%20src%3D%22jquery142.js%22%3E%3C/script%3E%0A");
		context.writeEscapedText("%09%09%3Cscript%20src%3D%22alligator.js%22%3E%3C/script%3E%0A");
		context.writeEscapedText("%09%09%3Cscript%3E%0A");
		context.writeEscapedText("%09%09%09function%20click1%28%29%7B%0A");
		context.writeEscapedText("%09%09%09%09server%28%22test%22%2Cfunction%28text%29%7B%0A");
		context.writeEscapedText("%09%09%09%09%09%24%28%22%23hello%22%29.html%28text%29%3B%0A");
		context.writeEscapedText("%09%09%09%09%7D%29%3B%0A");
		context.writeEscapedText("%09%09%09%7D%0A");
		context.writeEscapedText("%09%09%3C/script%3E%0A");
		context.writeEscapedText("%09%3C/head%3E%0A");
		context.writeEscapedText("%09%3Cbody%3E%0A");
		context.writeEscapedText("%09%09%3Cspan%20id%3D%22hello%22%3E%3C/span%3E%3Cbr/%3E%0A");
		context.writeEscapedText("%09%09%20%3Cform%3E%0A");
		context.writeEscapedText("%20%09%09%09%3Cinput%20type%3D%22button%22%20id%3D%22button1%22%20value%3D%22click%20me%22%20onclick%3D%22click1%28%29%22%20/%3E%0A");
		context.writeEscapedText("%20%09%09%3C/form%3E%0A");
		context.writeEscapedText("%09%3C/body%3E%0A");
		context.writeEscapedText("%3C/html%3E%0A");
	};
	
	this.test = function(context){
		context.write("hello ajax world");
	};


In case Memcached is enabled, the application and session contexts are being saved there.

# Configuration

Configuration is through a JSON text file - `settings.json` in
the same folder as `alligator.js`.

Example settings file:

	{
		"web_app_name" : "Alligator TestApp",
		"port"         : 8080,
		"path"         : {
					     	"root" : "/home/vadmin/ws/Alligator/WWW/",
					     	"lib"  : "/home/vadmin/ws/Alligator/WWWlib/"
				},
		"server_script": {
					     	 "template_ext"          : "jssp",
						     "script_ext"            : "ssjs",
						     "begin"                 : "<?",
						     "write" 				 : "=",
						     "global" 				 : "!",					     
						     "end"                   : "?>",
	                	     "session_minutes"       : 1,
					    	 "memcached"             : {
														   "enable" : 0,
														   "server" : "localhost",
														   "port"   : 11211
						                           		}
				         },
		"logs"		   :{
							"Engine"		:"DEBUG",
							"ASemaphore"	:"ERROR",
							"Session"		:"ERROR",
							"Application"	:"ERROR",
							"Page"			:"DEBUG"
						},
		"nodes"	       : 1
	}



This server listens on port 8080 for HTTP requests.

Explanation of properties:

- `web_app_name` - the name of the current web application
- `port` - the server listen to this port (default = 80)
- `path.root` - the root folder for the static and dynamic files (default = WWW)
- `path.lib` - the folder for JavaScript files that will get loaded automatically - you can access those via lib.fileName.yourStuff (default = None)
- `server_script.template_ext` - the extension of files that contains of server-side template (default = 'jssp')
- `server_script.script_ext` - the extension of files that contains of server-side script (default = 'ssjs')
- `server_script.begin` - the beginning tag for server-side scripting (default = '<?')
- `server_script.write` - in case coming right after the `server_script.begin` (e.g. '<?=') it auto warp the content with context.write()  (default = '=')
- `server_script.global` - in case coming right after the `server_script.begin` (e.g. '<?!') it adds the content outside the page() function  (default = '=')
- `server_script.end` - the end tag for server-side scripting (default = '?>')
- `server_script.session_minutes` - session timeout after X minutes (default = 30)
- `server_script.memcached` - integration with memcached 
- `logs` - the log level for each module
- `nodes` - number of process running this application server ** currently we support shared memory using memcached! we recommand to set the number of nodes as the number of cores

	
# Bugs and Contribution
Please let us know if you find any bug or if you would like to contribute code: mrohad.jsf at gmail

Known Bugs - http://github.com/mrohad/Alligator/blob/master/knowBugs.txt

# Credits

Original code forked from AntiNode @ http://github.com/mhansen/antinode
We are using Multi-node @ http://github.com/kriszyp/multi-node


ChangeLog
Change LOG 0.37
================================
Features:
1) Now compile into object whereas the main code is nested to the page() function
2) ability to write code outside the page() function using the <?!code?> tag
3) add examples to README
4) commands has been change to context
5) improvements to the logs module.


Change LOG 0.35
================================
Features:
1) Support script only file (SJS)
2) Performance improvements
	2.1) Cache Eval Functions
	2.2) prototyping

3) Support for NodeJS 0.2 has been verified



Change LOG 0.3
================================
Features:
1) Supports multi-node with session-scope and application-scope utilizing memcached.

BugFixes
1) bugs with static files support
 



Change LOG 0.29
================================
Features:
1) Compiling each JSSP file to JS file, upon a second use it evals/requires(the js file).

BugFixes
1) forward() is now working again



Change LOG 0.28
================================
Features:
1) Error details, on debug_mode Alligator generates a file by the name debugging.js, and add the error line to the response.





Change LOG 0.27
================================
Features:
1) multinode support - no shared memory yet

BugFixes
1) One can now write a server-side script with comments (//)


lib/engine.js
var EventEmitter = require('events').EventEmitter,
http = require('http'),
fs = require('fs'),
mime = require('./content-type'),
pathlib = require('path'),
uri = require('url'),
log = require('./log'),
sys = require('sys'),
Script = process.binding('evals').Script,
utils = require('./utils'),
sessionManager = require('./session'),
incomingForm = require('./incoming_form').incomingForm,
multi = require('./multinode/multi-node'),
asemaphore = require('./asemaphore'),
applicationManager = require('./application');
var moduleName = "Engine";
var cachedJssp = [];
var pageObjCache = new Object();
pageObjCache.get = function(key){
	var val = pageObjCache[key];
	return val;
};
pageObjCache.set = function(key,val){
	pageObjCache[key] = val;
};

var lib = new Object();
var globalSettings;
var applicationScope;
var sessionScope;

exports.start = function(settings) {
	globalSettings = defaultSettings(settings);
	var logObj = globalSettings.logs; 
	for (property in logObj)
		if(typeof logObj[property] != 'function')
			log.setLevel(property,logObj[property]); 
	log.info(moduleName,"Starting Web App - " +globalSettings.web_app_name);	
	//setting up sessionScope and applicationScope:
	applicationScope = applicationManager.start(globalSettings.web_app_name,globalSettings.server_script.memcached);
	sessionScope = sessionManager.start(globalSettings.server_script.session_minutes,globalSettings.web_app_name);
	//Load utils dynamicly 
	if(globalSettings.path.lib != undefined)
		fs.stat(globalSettings.path.lib, function (err, stats) {
			if (err) 
				log.error(moduleName,"Loading lib error:"+err);
			else if(!stats.isDirectory()) 
				log.error(moduleName,"Loading lib error: path must be a directory");
			else{
				var files = fs.readdirSync(globalSettings.path.lib);	
				for(i=0;i<files.length;i++){
					var file = files[i];
					if(files[i].endsWith(".js")){
						var fileNoJs =  file.substring(0,files[i].length-3);
						var path = pathlib.join(globalSettings.path.lib,fileNoJs);
						eval("lib."+fileNoJs+" = require(path)");
						log.info(moduleName,"Loading lib file:" + path);
					}
				}	
			}
		});
	//END util loading
	var server = http.createServer(function (req, res) {
		log.debug(moduleName,"request arrived!");
		var url = uri.parse(req.url,true);
		var pathname = (url.pathname || '/');
		var cleanPathname = pathname
		.replace(/\.\.\//g,'') //disallow parent directory access
		.replace(/\%20/g,' '); //convert spaces
		if(req.method == "GET"){
			var params = url.query;
			if(params == undefined)
				params = new Object();
			req.parameters = params;
			handleRequest(req,res,cleanPathname);			
		}
		else if (req.method == "POST"){
			incomingForm.parse(req, function(err, fields, files) {
				//log.debug(moduleName,"POST fields:" + utils.arrayToString(fields));
				params = new Object();
				req.parameters = fields;
				handleRequest(req,res,cleanPathname);
			});
		}
		else //Other Methods
			handleRequest(req,res,cleanPathname);


	});
	multi.listen({
		port: globalSettings.port, 
		nodes: globalSettings.nodes
	}, server);

	log.warn(moduleName,'Server running at port '+globalSettings.port);
};

function handleRequest(req,res,cleanPathname,newSessionId){
	var root = globalSettings.path.root;
	var path = pathlib.join(root, cleanPathname);
	if(newSessionId==undefined)
		log.info(moduleName,"Handling request to: " +path + " pid("+process.pid+")");
	else
		log.info(moduleName,"Forwarding request to: " +path + " pid("+process.pid+")");
	//log.debug(moduleName,"Request headers: "+utils.arrayToString(req.headers));
	fs.stat(path, function (err, stats) {
		if (err) {
			// ENOENT is normal on 'file not found'
			if (err.errno != process.ENOENT) { 
				// any other error is abnormal - log it
				log.error("fs.stat(",path,") failed: ", err);
			}
			return fileNotFound(req,res,path);
		}
		if (!stats.isFile() && !stats.isDirectory()) 
			return fileNotFound(req,res,path);
		else{
			if (stats.isDirectory()) 
				path = pathlib.join(path, "index.html");
			var cookie = req.headers["cookie"];
			var sessionId = utils.getSessionId(cookie);
			if(newSessionId!=undefined)//forward
				sessionId = newSessionId;
			else{
				sessionScope.hello(sessionId);
			}
			if(!(path.endsWith("."+globalSettings.server_script.template_ext) || path.endsWith("."+globalSettings.server_script.script_ext))){
				log.info(moduleName,"Static request");
				sendHeaders(req, res,undefined, stats.size, mime.mimeType(path), stats.mtime); 
				var readStream = fs.createReadStream(path);
				sys.pump(readStream,res);					
			}else{
				log.info(moduleName,"Dyanmic request");
				if(path.endsWith("."+globalSettings.server_script.template_ext) && (cachedJssp[path] == undefined || cachedJssp[path+"#date"] != stats.mtime.toUTCString())){
					var readStream = fs.createReadStream(path);
					var script = [];
					readStream.addListener("data", function (chunk) {	
						script.push(chunk.toString());	
					});
					readStream.addListener("end", function () {
						log.info(moduleName,"STARTING PROCESSING JSSP");		
						SSJSWriter(script.join(""),req,res,path,stats.mtime,sessionId);
						log.info(moduleName,"END OF JSSP PROCESSING");		
					});
					req.connection.addListener('timeout', function() {
						/* dont destroy it when the fd's already closed */
						if (readStream.readable) {
							log.debug(moduleName,'timed out. destroying file read stream');
							readStream.destroy();
						}
					});
					res.addListener('error', function (err) {
						log.error('error writing',file,sys.inspect(err));
						readStream.destroy();
					});
					readStream.addListener('fd', function(fd) {
						log.debug(moduleName,"opened",path,"on fd",fd);
					});

					readStream.addListener('error', function (err) {
						log.error('error reading',file,sys.inspect(err));
						resp.end('');
					});

				}
				else{
					log.info(moduleName,"RUN JSSP FROM CACHE");
					var currentFileNameToRead = cachedJssp[path];
					if(path.endsWith("."+globalSettings.server_script.script_ext))
						currentFileNameToRead = path.substring(0,path.length-globalSettings.server_script.script_ext.length-1);;
					serverSideRunning(currentFileNameToRead,req,res,path,stats.mtime,sessionId,false);
				}

			}

		}		

	});
};

function serverSideRunning(newfileName,request,response,file,lastMod,sessionId,isForceEval){
	var responseHead= new Object();	
	var result = new Object();
	result.html = "";
	var flushResponse = true;
	var flushFunction = undefined;
	var error = false;
	responseHead.status = 200;
	responseHead.headers = {};
	var afterEval = [];
	//log.debug(moduleName,asemaphore);
	var currentAsemaphore = asemaphore.ctor(1,function(){
		if(!error)
			result.html = afterEval.join("");
		if(flushResponse){//otherwise, forwarding...
			sendHeaders(request,response,responseHead,result.html.length,mime.mimeType(file,"text/html"),lastMod,result.sessionId); 
			response.end(result.html);
		}
		else if(flushFunction)
			flashFunction();
	});

	var executeTheSJSFile = function (errRead, functoRun) {
		try{
			if (errRead) throw errRead;
			//Handling session and application with asemphore
			result.sessionId = sessionId;
			var application = applicationManager.getManager("",currentAsemaphore);
			var session = sessionManager.getManager(sessionId,currentAsemaphore,applicationManager,result);
			//-----------
			//Genrating Context
			var context = {
					request:request,
					responseHead:responseHead,
					application:application,
					session:session,
					lib:lib,
					log:log,
					write : function(text){
						//log.debug(moduleName,"WRITE afterEval : "+ afterEval);
						afterEval.push(text);
					},
					writeEscapedText :function(text){
						afterEval.push(unescape(text));
					},
					forward :function(resource){
						flushFunction = handleRequest(request,response,resource,result.sessionId);
						flushResponse = false;
					},
					sendRedirect:function(url){
						responseHead.status = 301;
						responseHead.headers["location"] = url;
					}
				};
			//=============================================
			//log.debug(moduleName,typeof functoRun.page);
			log.debug(moduleName,"READING pageObject from cache, key - "+file );
			var pageObjFromCache = pageObjCache.get(file);
			if(pageObjFromCache == undefined || isForceEval){//not in cache	
				log.debug(moduleName,"....and....cache is EMPTY for key - "+file);
				var pageObjFunc = new Function(functoRun.toString() + " ; return this;");
				pageObjFromCache = new pageObjFunc(context);
				pageObjCache.set(file,pageObjFromCache);
			}
			else
				log.debug(moduleName,"....and....cache is FULL for key - "+file);
			if(request.parameters.serverFunction == undefined){
				pageObjFromCache.page(context);
			}else{//Server command
				serverFunction = request.parameters.serverFunction;
				pageObjFromCache[serverFunction](context);
			} 
			currentAsemaphore.p();
			//log.debug(moduleName,"sessionId - "+result.sessionId);
		}
		catch(err){
			log.error("parse problem:"+err+" stack:"+err.stack);
			flushResponse = true;
			error = true;
			responseHead.status = 500;
			result.html = ["<h1>",globalSettings.web_app_name," - SERVER ERROR</h1>."].join("");
			if(log.getLevel(moduleName)==0){
				result.html += "(Debug Mode) Could not parse jssp<br/> ";
				var erStr = "Details: ";
				result.html += erStr+err.stack;
			}
			else
				result.html += "Could not parse jssp";
			while(currentAsemaphore.p()>0);
		}
	};
	
	if(cachedJssp[file]!= undefined && !isForceEval){//funcCaching
		log.debug(moduleName,"SSJS RUN: bring file from CACHE:  "+file);
		executeTheSJSFile(undefined,undefined);
	}else{
		var fileNameToRead = [newfileName,".",globalSettings.server_script.script_ext].join("");
		log.debug(moduleName,"SSJS RUN: reading file from FS:  "+fileNameToRead);
		fs.readFile(fileNameToRead,executeTheSJSFile);
	}
}

function SSJSWriter(str,request,response,file,lastMod,sessionId){
	var parsedTextArray = [];
	var externalParsedTextArray = [];
	var startTag = globalSettings.server_script.begin;
	var startWriteAddition = globalSettings.server_script.write;
	var startGlobalAddition = globalSettings.server_script.global;
	var endTag = globalSettings.server_script.end;
	var lineArray = str.split(new RegExp( "\\n", "g" ));
	var isInScript = false;
	var currentScript =[];
	var nextLine = "\n";
	for(index=0;index<lineArray.length;index++){
		line = lineArray[index];			
		while(line.length>0){
			if(!isInScript){
				var startTagIndex = line.indexOf(startTag);
				if(line.indexOf(startTag)==-1){
					parsedTextArray.push('context.writeEscapedText("'+escape(line+nextLine)+'");'+nextLine);
					line="";
				}
				else{
					lineBeforeStart = line.substring(0,startTagIndex);
					parsedTextArray.push('context.writeEscapedText("'+escape(lineBeforeStart)+'");');
					line = line.substring(startTagIndex+startTag.length);
					if(line.length==0)
						parsedTextArray.push(nextLine);
					isInScript = true;
				}
			}
			else{//Inscript
				var endTagIndex =line.indexOf(endTag);
				if(line.indexOf(endTag)==-1){
					currentScript.push(line+nextLine);
					line="";
				}
				else{
					lineBeforeEnd = line.substring(0,endTagIndex);
					currentScript.push(lineBeforeEnd);
					var theScript = currentScript.join("");
					if(theScript.startsWith(startWriteAddition)){ //handling <?=...?> cases
						theScript = "context.write("+theScript.substring(startWriteAddition.length)+");";
						parsedTextArray.push(theScript);
					}else if(theScript.startsWith(startGlobalAddition)){ //handling <?!...?> cases
						theScript = theScript.substring(startGlobalAddition.length);
						externalParsedTextArray.push(theScript);
					}else{
						parsedTextArray.push(theScript);
					}
					currentScript = [];
					line = line.substring(endTagIndex+endTag.length);
					if(line.length==0)
						parsedTextArray.push(nextLine);
					isInScript = false;
				}
			}
		}
	}
	//var toEval =
	//Waiting for a bug fix(V8)
	//http://groups.google.com/group/nodejs/browse_thread/thread/7a2e409ec970198e/d9336b7b2764f129?lnk=gst&q=require+exception#d9336b7b2764f129
	//var finalFunction = "exports.run = (function(lib,application,request,responseHead,writeEscapedText,forward,sendRedirect,write,session) {"
	//		+toEval+"})";
	var finalFunction = ["this.page = function (context){",parsedTextArray.join(""),"};",externalParsedTextArray.join("")].join("");
	var newfileName = file.substring(0,file.length-globalSettings.server_script.template_ext.length-1);
	fs.writeFile([newfileName,".",globalSettings.server_script.script_ext].join(""), finalFunction, function (err) {
		//log.debug(moduleName,"Error writin cache file: "+err); 
		cachedJssp[file] = newfileName;
		cachedJssp[file + "#date"] = lastMod.toUTCString();
		log.info(moduleName,"Caching  - "+file + ",last mod - "+ lastMod.toUTCString());
		serverSideRunning(newfileName,request,response,file,lastMod,sessionId,true);
	});

}

function sendHeaders(req, res,responseHead, length, content_type, modified_time,sessionId) {
	//log.debug(moduleName,"send headers: sessionId - "+sessionId);
	if(responseHead==undefined)
		responseHead = new Object();
	if(responseHead.status==undefined)
		responseHead.status = 200;
	if(responseHead.headers==undefined)	
		responseHead.headers = {};
	responseHead.headers["date"] = (new Date()).toUTCString();
	responseHead.headers["Server"] = "Alligator/0.3+ Node.js/"+process.version;
	if(sessionId != undefined)
		if(responseHead.headers["Set-cookie"] == undefined)//TODO add expiary and domain
			responseHead.headers["Set-cookie"] = "njssession="+sessionId;
		else
			responseHead.headers["Set-cookie"] += ";njssession="+sessionId;
	if (length) 
		responseHead.headers["Content-Length"] = length;
	if (content_type) 
		responseHead.headers["Content-Type"] = content_type || "application/octet-stream";
	if (modified_time) 
		responseHead.headers["Last-Modified"] = modified_time.toUTCString(); 
	//log.debug(moduleName,"RESPONSE Headers :"+utils.arrayToString(responseHead.headers)+" +++  RESPONSE Status :"+responseHead.status);
	res.writeHead(responseHead.status, responseHead.headers);
	log.info(moduleName,req.connection.remoteAddress,req.method,responseHead.status,length);
}

function fileNotFound(req,res,path) {
	log.debug(moduleName,"404 opening path: '"+path+"'");
	var body = "404: " + req.url + " not found.\n";
	var responseHead= new Object();
	responseHead.status = 404;
	sendHeaders(req, res,responseHead,body.length,"text/plain");
	if (req.method != 'HEAD') 
		res.end(body, 'utf-8');
	else 
		res.end('');
}

function defaultSettings(settings){
	if(settings.web_app_name == undefined)
		settings.web_app_name = "AlligatorWebApp ";
	if(settings.port == undefined)
		settings.port = 80;
	if(settings.path == undefined)
		settings.path = new Object();
	if(settings.path.root == undefined)
		settings.path.root = "WWW";	
	if(settings.server_script == undefined)
		settings.server_script = new Object();
	if(settings.server_script.temlate_ext == undefined)
		settings.server_script.temlate_ext = "jssp";
	if(settings.server_script.script_ext == undefined)
		settings.server_script.script_ext = "ssjs";
	if(settings.server_script.begin == undefined)
		settings.server_script.begin = "<?";
	if(settings.server_script.end == undefined)
		settings.server_script.end = "?>";
	if(settings.server_script.write == undefined)
		settings.server_script.write = "=";
	if(settings.server_script.global == undefined)
		settings.server_script.global = "!";
	//memcached
	if(settings.server_script.memcached == undefined)
		settings.server_script.memcached = new Object();
	if(settings.server_script.memcached.enable == undefined)
		settings.server_script.memcached.enable = 0;
	if(settings.server_script.memcached.server == undefined)
		settings.server_script.memcached.server = "localhost";
	if(settings.server_script.memcached.port == undefined)
		settings.server_script.memcached.port = 11211;

	if(settings.logs == undefined)
		settings.logs = new Object();
	if(settings.nodes == undefined)
		settings.nodes = 1;
	return settings;
}

